### PR TITLE
move sample realtime events for put/patch/delete to api layer

### PR DESCRIPTION
**Pull request by shriramshankar**

### CODE REVIEW

**Comment by shriramshankar** [comment]
@annyhe @iamigo PR #292 needs to be merged before this one

File: realtime/redisPublisher.js [hunk 1]
```diff
@@ -92,32 +92,49 @@ function publishObject(inst, event, changedKeys, ignoreAttributes) {
...
-  const subName = sample.name.split('|')[0];
-  const options = {};
-  options.where = { absolutePath: subName };
-  return model.findOne(options)
+  const nameParts = sample.name.split('|');
+  const subName = nameParts[0];
+  const aspName = nameParts[1];
+  const subOpts = {
+    where: {
+      absolutePath: subName,
```
Comment (iamigo) [inline realtime/redisPublisher.js]: case insensitive?

File: realtime/redisPublisher.js [hunk 2]
```diff
@@ -92,32 +92,49 @@ function publishObject(inst, event, changedKeys, ignoreAttributes) {
...
+    },
+  };
+  const aspOpts = {
+    where: {
+      name: aspName,
```
Comment (iamigo) [inline realtime/redisPublisher.js]: case insensitive?

Comment (shriramshankar) [inline realtime/redisPublisher.js]: aspectName and subjectName are always from the "sample.name" field from the db and the "sample.name" field is formed from its subject and its aspect in the db. So the case always matches when we do the find here.

File: tests/realtime/redisPublisher.js [hunk 3]
```diff
@@ -81,6 +81,29 @@ describe('redis Publisher', () => {
       })
       .catch(done);
     });
+
+    it('when tried to publish sample without aspect,'+
```
Comment (iamigo) [inline tests/realtime/redisPublisher.js]: any other tests we can add here to make sure that we aren't breaking any of the existing realtime event publishing?

Comment (shriramshankar) [inline tests/realtime/redisPublisher.js]: The sample object the is finally passed to the real-time emitter should have the following fields/objects to not to break anything
1. absolutePath - as a field copied over from subject absolutePath
2. subject object always attached - having  tags field
3. aspect object always attached having tags field
I have added tests to check for these.

File: api/v1/helpers/verbs/doPut.js [hunk 4]
```diff
@@ -75,6 +77,14 @@ function doPut(req, res, next, props) {
   putPromise.then((o) => {
     resultObj.dbTime = new Date() - resultObj.reqStartTime;
     u.logAPI(req, resultObj, o);
+
+    // publish the update event to the redis channel
+    if (props.publishEvents) {
+      publisher.publishSample(o, props.associatedModels.subject,
+
```
Comment (annyhe) [inline api/v1/helpers/verbs/doPut.js]: is there a reason we are adding an empty line here?

File: db/model/sample.js [hunk 5]
```diff
@@ -320,52 +322,6 @@ module.exports = function sample(seq, dataTypes) {
       }, // hooks.beforeCreate
 
       /**
-       * Publishes the created sample to redis channel *including* the values
-       * from its aspect association.
-       *
-       * @param {Sample} inst - The newly-created instance
-       */
-      afterCreate(inst /* , opts */) {
-        let samp;
```
Comment (annyhe) [inline db/model/sample.js]: can we test for whether a sample is published on realtime create and delete?

Comment (shriramshankar) [inline db/model/sample.js]: There is another story for this.

**Comment by shriramshankar** [comment]
@iamigo @annyhe  Changes done and PR updated.

**Comment by shriramshankar** [comment]
@annyhe  Can you merge this PR?

**Pull request merged**